examples/models/custom_metrics/ModelWithMetrics.py

```diff
@@ -0,0 +1,16 @@
+class ModelWithMetrics(object):
+
+    def __init__(self):
+        print("Initialising")
+
+    def predict(self,X,features_names):
+        print("Predict called")
+        return X
+
+    def metrics(self):
```

Comment:
If I understand correctly, this is how to define custom metrics in the new wrapper ?
In the past we send custom metrics as part of the predict response then the engine will pick it up and aggregate it.

Will the old way still working in this release or we need to switch to this ?

Also I am wondering what's the suggest way of providing a request level metrics, e.g. predict_latency, as this is request level, I am not sure what's the best way to share data between `predict` method and `metrics` method
Should we define our own prometheus metrics then expose it through `metrics` method ?

Reply:
Just to clarify all the standard metrics are were available via engine are still available via executor and works out of the box - this includes latency, requests per second, etc.
Regarding `predict_latency` - I believe similar metric is gathered by the executor now.

Currently whatever is returned by `metrics` function in Python wrapper had been already added to the response automatically and, as you said, picked up by the `engine`.

As `executor` is not inspecting payloads (to optimise for performance) it cannot aggregate these metrics and therefore we expose them directly from the wrapper. The new endpoint that serves metrics to be scraped by Prometheus is taking them only from the `metrics` function defined in the model class (also handled in multi worker / multi thread case).

Are you by any chance using `predict_raw` instead of  `predict`?

Could you provide us with a minimal example on how you collect metrics at the moment?
We will greatly appreciate feedback as it is key on how we collect custom metrics.

Reply:
Just as a heads up we are looking to make `metrics` function as a standard way of providing custom metrics - does this work with your current workflow?

Reply:
Right, we are using `predict_raw` and the metrics are sent back through response `
`meta` field.

For our usage, we also want to measure other time cost during `predict_raw`, e.g. fetching data from other services etc. It looks like with this change, we should use Prometheus library in our codes, then records metrics by it and expose data from `metrics` function.

The part I don't know how to deal with it properly is metrics type `TIMER`, that's not a standard prometheus metrics.

Are you saying when Prometheus scrape the endpoint, it actually just invokes `/metrics` to get data ?

Reply:
> The part I don't know how to deal with it properly is metrics type TIMER, that's not a standard prometheus metrics.

My understanding is that `TIMER` is what is known as `HISTOGRAM` in `Prometheus`. 

> Are you saying when Prometheus scrape the endpoint, it actually just invokes /metrics to get data ?

Yes, exactly. And the actual endpoint will actually be called /prometheus (we configure that via environmental variables). 


Thanks for the info about usage of `predict_raw`. Do I understand correctly that using metrics returned from it instead of one from `metrics` method would address your use case?

Reply:
If you could provide us with a minimal mock model that has the same structure as your use case it would be a great help!

Reply:
Thanks for the info about usage of predict_raw. Do I understand correctly that using metrics returned from it instead of one from metrics method would address your use case?

> Yes, that's what we are doing now

If you could provide us with a minimal mock model that has the same structure as your use case it would be a great help!

> I attached a mock model similar to what we did, I think my main question is how to deal with `TIMER` type, in the past we just return the time spent at request level and relying on seldon-engine did the aggregation and conversion to `HISTORGAM` in `Prometheus`.
with the new `metrics` method I am not sure what value should I return for these `TIMER`  type metrics.

```
class BaseModel:

    def predict_raw(self, msg: dict) -> dict:
    
        data_fetch_start = time.time()
        model_input = fetch_data()
        data_fetch_time = since_time(data_fetch_start)
        metrics.append(Metric(MetricType.FT, data_fetch_time))

        predict_start = time.time()
        model_output = self.model.predict(model_input)
        predict_time = since_time(predict_start)
        metrics.append(Metric(MetricType.PT, predict_time))

        ndarray = model_output.to_ndarray()

        prometheus_metrics = [
            {
                "type": m.metric_type.name,
                "key": m.metric_type.key,
                "value": m.value,
            } for m in metrics
        ]

        return {
            "meta": {"metrics": prometheus_metrics},
            "data": {"names": columns, "ndarray": ndarray},
        }
```

Reply:
Thanks for the example, we will make sure to make it still possible to use `predict_raw` and its custom metrics reaching `Prometheus`.

The new behaviour is exactly the same as it was before: code of new `SeldonMetrics` class reproduces behaviour that was present in the engine -> hence the logic for constructing the histogram.

Reply:
Got it, thanks!